### PR TITLE
NUTCH-2038

### PR DESCRIPTION
**Pull request by asitang**
repulled

### CODE REVIEW

File: ivy/ivy.xml [hunk 1]
```diff
@@ -78,7 +78,11 @@
                 <dependency org="org.apache.cxf" name="cxf-rt-transports-http-jetty" rev="3.0.4"/>
                 <dependency org="com.fasterxml.jackson.core" name="jackson-databind" rev="2.5.1" /> 
                 <dependency org="com.fasterxml.jackson.dataformat" name="jackson-dataformat-cbor" rev="2.5.1" />
-                <dependency org="com.fasterxml.jackson.jaxrs" name="jackson-jaxrs-json-provider" rev="2.5.1" />	
+                <dependency org="com.fasterxml.jackson.jaxrs" name="jackson-jaxrs-json-provider" rev="2.5.1" />
+                <dependency org="org.apache.mahout" name="mahout-math" rev="0.8" />
```
Comment (lewismc) [inline ivy/ivy.xml]: The Mahout and Lucene dependencies cannot be included in main ivy.xml if this is to be a plugin.

File: src/java/org/apache/nutch/parse/ModelURLFilterAbstract.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package org.apache.nutch.parse;
```
Comment (lewismc) [inline src/java/org/apache/nutch/parse/ModelURLFilterAbstract.java]: We need license headers

File: conf/nutch-default.xml [hunk 3]
```diff
@@ -1136,6 +1136,28 @@
 <!-- parser properties -->
 
 <property>
+  <name>parser.modelfilter.trainfile</name>
+  <value>tweets-train.tsv</value>
+  <description>
```
Comment (lewismc) [inline conf/nutch-default.xml]: These all need descriptions. If they do not have descriptions then no one has a clue what they do.

File: src/java/org/apache/nutch/net/URLFilters.java [hunk 4]
```diff
@@ -41,4 +41,24 @@ public String filter(String urlString) throws URLFilterException {
     }
     return urlString;
   }
+/**Get a filter with the full classname if only it is activated through the nutchsite.xml*/
```
Comment (lewismc) [inline src/java/org/apache/nutch/net/URLFilters.java]: This is messy Javadoc. You should make efforts to reference the configuration file properly e.g. nutch-site.xml as well as include the method parameter description(s).

File: src/java/org/apache/nutch/parse/ParseSegment.java [hunk 5]
```diff
@@ -56,6 +57,14 @@
   private ParseUtil parseUtil;
 
   private boolean skipTruncated;
+  
+  public static final String PARSER_MODELFILTER="parser.modelfilter";
```
Comment (lewismc) [inline src/java/org/apache/nutch/parse/ParseSegment.java]: I would argue that the key's are unsuitable. As this is meant to be an URLFilter, the keys should refelect that... not parser related but instead URLFilter specific

File: src/java/org/apache/nutch/parse/ParseSegment.java [hunk 6]
```diff
@@ -140,6 +161,29 @@ public void map(WritableComparable<?> key, Content content,
           LOG.warn("Error passing score: " + url + ": " + e.getMessage());
         }
       }
+      
+if(filterflag){
+          
+
```
Comment (lewismc) [inline src/java/org/apache/nutch/parse/ParseSegment.java]: Code formatting is all over the place here. We have 2 space indents.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/ModelURLFilter.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+import java.io.BufferedReader;
+
+import java.io.IOException;
+
+import java.util.ArrayList;
+
+/**
+ * Filters URLs based on a file of URL prefixes. The file is named by (1)
+ * property "urlfilter.prefix.file" in ./conf/nutch-default.xml, and (2)
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/ModelURLFilter.java]: Is this Javadoc correct? It doesn't seem to agree with the proposed changes to nutch-default.xml

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/ModelURLFilter.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ * attribute "file" in plugin.xml of this plugin Attribute "file" has higher
+ * precedence if defined.
+ * 
+ * <p>
+ * The format of this file is one URL prefix per line.
+ * </p>
+ */
+public class ModelURLFilter extends ModelURLFilterAbstract {
+
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/ModelURLFilter.java]: Your formatting is all over the place. We have 2 space indents in the Nutch codebase.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/ModelURLFilter.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	private boolean relevent = false;
+	private Configuration conf;
+	private String inputFilePath;
+	private String dictionaryFile;
+	private ArrayList<String> wordlist = new ArrayList<String>();
+
+	public ModelURLFilter() throws Exception {
+
+	}
+
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/ModelURLFilter.java]: None of the methods in this class are documented... this is not ideal as it makes it difficult for other to understand what is going on.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java [hunk 10]
```diff
@@ -0,0 +1,234 @@
+/**
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java]: The name of this file should be NaiveBayesClassifier as the NBClassifier is highly ambiguous

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+import org.apache.mahout.math.RandomAccessSparseVector;
+import org.apache.mahout.math.Vector;
+import org.apache.mahout.math.Vector.Element;
+import org.apache.mahout.vectorizer.SparseVectorsFromSequenceFiles;
+import org.apache.mahout.vectorizer.TFIDF;
+
+import com.google.common.collect.ConcurrentHashMultiset;
+import com.google.common.collect.Multiset;
+
+public class NBClassifier {
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java]: This class and all of it;s methods need to be documented. No-one will know what is going on here.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+	public static Map<String, Integer> readDictionnary(Configuration conf,
+			Path dictionnaryPath) {
+		Map<String, Integer> dictionnary = new HashMap<String, Integer>();
+		for (Pair<Text, IntWritable> pair : new SequenceFileIterable<Text, IntWritable>(
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java]: Formatting in this file is all over the place, in Nutch we use 2 space indents.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+			value.set(message);
+			writer.append(key, value);
+
+		}
+		reader.close();
+		writer.close();
+
+	}
+
+	public static void main(String args[]) throws Exception {
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java]: Please implement or remove the commented out code.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/package-info.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * URL filter plugin to include only URLs which match an element in a given list of
+ * domain suffixes, domain names, and/or host names.
+ * See {@link org.apache.nutch.urlfilter.domainblacklist} for the counterpart
```
Comment (lewismc) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/package-info.java]: Is this correct? The Javadoc seems a bit strange

**Comment by asitang** [comment]
Pulled

**Comment by asitang** [comment]
sorry for the mess reopened a new one

File: ivy/ivy.xml (hunk 1 above)
Comment (asitang) [inline ivy/ivy.xml]: Was trying to pave the way for a machine learning library into nutch, so that anyone can use that in future

Comment (lewismc) [inline ivy/ivy.xml]: Hi Asitang,
I get Your point however I am also trying OO help you get your path into
the codebase.  Nutch is a crawler... Adding machine learning  and indexing
components such as Mahout (what if someone does not wish to use Mahout) and
Lucene (what if someone wishes to use ES) back into the core codebase
dependency tree is, on this occasion,, not the right way to go.
If you can ease send a new pull request we can take a look. Excellent work
thank you :)

On Thursday, June 18, 2015, asitang <notifications@github.com> wrote:

> In ivy/ivy.xml
> <https://github.com/apache/nutch/pull/32#discussion_r32741196>:
>
> > @@ -78,7 +78,11 @@
> >                  <dependency org="org.apache.cxf" name="cxf-rt-transports-http-jetty" rev="3.0.4"/>
> >                  <dependency org="com.fasterxml.jackson.core" name="jackson-databind" rev="2.5.1" />
> >                  <dependency org="com.fasterxml.jackson.dataformat" name="jackson-dataformat-cbor" rev="2.5.1" />
> > -                <dependency org="com.fasterxml.jackson.jaxrs" name="jackson-jaxrs-json-provider" rev="2.5.1" />	
> > +                <dependency org="com.fasterxml.jackson.jaxrs" name="jackson-jaxrs-json-provider" rev="2.5.1" />
> > +                <dependency org="org.apache.mahout" name="mahout-math" rev="0.8" />
>
> Was trying to pave the way for a machine learning library into nutch, so
> that anyone can use that in future
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/nutch/pull/32/files#r32741196>.
>


-- 
*Lewis*

File: conf/nutch-default.xml (hunk 3 above)
Comment (asitang) [inline conf/nutch-default.xml]: I will update the code for all the formatting and info changes. I had already mentioned it in jira. This is an "Uncouth" patch just to show the flow and the proof of concept. Thanks for the formatting pointers though :)

File: ivy/ivy.xml (hunk 1 above)
Comment (chrismattmann) [inline ivy/ivy.xml]: why not just put this in the plugin's ivy.xml, we can do that Asitang, and Lewis, right? @lewismc @asitang

File: src/java/org/apache/nutch/parse/ParseSegment.java (hunk 5 above)
Comment (chrismattmann) [inline src/java/org/apache/nutch/parse/ParseSegment.java]: yeah should be insulated to the plugin.

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java (hunk 10 above)
Comment (chrismattmann) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java]: +1

File: src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java (hunk 13 above)
Comment (chrismattmann) [inline src/plugin/urlfilter-model/src/java/org/apache/nutch/urlfilter/model/NBClassifier.java]: +1

**Comment by chrismattmann** [comment]
hi @asitang please see my comments

**Pull request closed without merging**